File: benchmarks/tasks/either_validate/either_validate.hs
-- SPDX-License-Identifier: MIT-0
module Main where

data Result a b = Err a | Ok b

validateName :: String -> Result String String
validateName "" = Err "empty name"
validateName n  = Ok n

validateAge :: Int -> Result String Int
validateAge a
  | a > 0 && a < 150 = Ok a
  | otherwise         = Err "invalid age"

showResult :: Result String String -> String
showResult (Ok s)  = "Ok " ++ s
showResult (Err s) = "Err " ++ s

main :: IO ()
main = do
  putStrLn (showResult (validate "Alice" 25))
  putStrLn (showResult (validate "" 25))
  where
    validate name age =
      case validateName name of
        Err e -> Err e
        Ok n  -> case validateAge age of
          Err e -> Err e
          Ok a  -> Ok (n ++ " is " ++ show a)

File: benchmarks/tasks/either_validate/either_validate.py
# SPDX-License-Identifier: MIT-0

def validate_name(name):
    if name == "":
        return ("err", "empty name")
    return ("ok", name)

def validate_age(age):
    if age < 0:
        return ("err", "negative age")
    return ("ok", age)

def and_then(result, f):
    if result[0] == "err":
        return result
    return f(result[1])

def validate(name, age):
    return and_then(validate_name(name),
        lambda n: and_then(validate_age(age),
            lambda a: ("ok", f"{n} is {a}")))

def show_result(r):
    if r[0] == "ok":
        return f"Ok {r[1]}"
    return f"Err {r[1]}"

print(show_result(validate("Alice", 25)))
print(show_result(validate("", -1)))
